src/coreclr/jit/llvm.cpp

```diff
@@ -248,6 +359,56 @@ void castingStore(llvm::IRBuilder<>& builder, Value* toStore, Value* address, va
     castingStore(builder, toStore, address, getLLVMTypeForVarType(type));
 }
 
+/// <summary>
+/// Returns the llvm arg number or shadow stack offset for the corresponding local which must be loaded from an argument
+/// </summary>
+struct LlvmArgInfo getLlvmArgInfoForArgIx(CORINFO_SIG_INFO& sigInfo, unsigned int lclNum)
+{
+    if (sigInfo.hasExplicitThis() || sigInfo.hasTypeArg())
+        failFunctionCompilation();
+
+    unsigned int llvmArgNum    = 1; // skip shadow stack arg
+    bool         returnOnStack = false;
+
+    if (needsReturnStackSlot(sigInfo.retType))
+    {
+        llvmArgNum++;
+    }
+
+    CORINFO_ARG_LIST_HANDLE sigArgs     = sigInfo.args;
+    struct LlvmArgInfo      llvmArgInfo = {
+        -1 /* default to not an LLVM arg*/, sigInfo.hasThis() ? TARGET_POINTER_SIZE : 0 /* this is the first pointer on
+                                                                                           the shadow stack */
+    };
+    unsigned int shadowStackOffset = llvmArgInfo.m_shadowStackOffset;
+
+    for (unsigned int i = 0; i < sigInfo.numArgs; i++, sigArgs = _info.compCompHnd->getArgNext(sigArgs))
+    {
+        CorInfoType corInfoType = getCorInfoTypeForArg(sigInfo, sigArgs);
+        if (canStoreTypeOnLlvmStack(corInfoType))
+        {
+            if (lclNum == i)
+            {
+                llvmArgInfo.m_argIx = llvmArgNum;
+                return llvmArgInfo;
+            }
+
+            llvmArgNum++;
+        }
+        else
+        {
+            if (lclNum == i)
+            {
+                llvmArgInfo.m_shadowStackOffset = shadowStackOffset;
+                return llvmArgInfo;
+            }
+
+            shadowStackOffset += TARGET_POINTER_SIZE; // TODO size of arg, for now only handles byrefs and class types
+        }
+    }
+    failFunctionCompilation(); // lclNum not an argument, TODO: how to indicate a failure in compilation
```

Comment:
Failure from bad, like bad IL? I suppose, the frontend should've failed on it. Though here it looks like you would like to assert the path is unreachable?

We have an `unreachable()` macro, or you can `assert(!"lclNum not an argument");` (more informative in my opinion). There is also a `noway_assert`, which is an `assert` that's enabled in Release builds (for our purposes, in CoreCLR it also makes the EE try to compile the function in MinOpts).

Reply:
I guess I was thinking, more failure in my own programming skills, and hence a bug in my code.  I'll go with `assert`.

Reply:
followed by `UNREACHABLE()`

Reply:
FWIW, `unreachable()` is usually used as an escape hatch from compiler warnings, or for truly "unreachable" stuff. I would think an assert is enough?

Reply:
One thing to note is that `unreachable()` is on in Release builds, but does that matter given `gtTreeID` is `DEBUG`-only, so Release builds are not that big a concern?

Reply:
"gtTreeID is DEBUG-only"  Oh that's a shame.  We used to support Release builds in the CoreRT repo.  It got dropped in the move to runtimelab, but I would like to enable them again.  I'd better look at removing that in the next PR.  For this change, I'll need to return something, or call something that never returns.  I'll add a `failFunctionCompilation` so the compiler is ok with no return value.  That will also be ok for Release builds while the transition from IL to IR is in progress.

Reply:
> 'll need to return something, or call something that never returns

Oh I see. Then `unreachable()` is the macro to use.

Reply:
At the risk of continuing the discussion, I replaced the `for` with  a `do/while` to remove the multiple returns and while still a bit of an odd structure, I think it is a small improvement.